### PR TITLE
design: thin pool

### PR DESCRIPTION
**Pull request by sp98**
Signed-off-by: Santosh Pillai <sapillai@redhat.com>

### CODE REVIEW

File: doc/design/thin_pool.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+## Summary
+Allow LVM operator to create thin-pools along with the volume groups.
+
+
+## Proposal:
+- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a volume group and a thin-pool.
+- One volume group will be mapped to a single thin pool.
+- User should be able to configure the thin-pool size based on percentage of the available volume group size.
+- Default chunk size of the thin pool should be 512 kiB
```
Comment (nbalacha) [inline doc/design/thin_pool.md:11]:
```suggestion
- Default chunk size of the thin pool will be 512 kiB
```

File: doc/design/thin_pool.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
++   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+}
+```
+
+
+- Following new fields will added to `DeviceClass` API
+    - **ThinPoolConfig** API contains information related to a thin pool.These configuration options are:
+        - **Name**: Name of the thin-pool
+        - **SizePercent**: Size of the thin pool to be created with respect to available free space in the volume group. It represents percentage value and not absolute size values. Size value should range between 10-90. It defaults to 75 if no value is provided.
+        - **OverprovisionRatio**: OverprovisionRatio represents the ratio of overprovision that can be allowed on thin pools.
```
Comment (nbalacha) [inline doc/design/thin_pool.md:56]:
```suggestion
        - **OverprovisionRatio**: The factor by which additional storage can be provisioned compared to the available storage in the thin pool.
```

File: doc/design/thin_pool.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+- `LVMVolumeGroup` API changes:
+
+``` go=
+type LVMVolumeGroupSpec struct {
+    // DeviceSelector is a set of rules that should match for a device to be
+    // included in this TopoLVMCluster
```
Comment (nbalacha) [inline doc/design/thin_pool.md:63]:
```suggestion
    // included in this VolumeGroup
```

File: doc/design/thin_pool.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+- Volume Group Manager is responsible for creating the thin-pools after creating the volume group.
+- Command used for creating the thin-pool:
+    ```
+    lvcreate -L <Size>%FREE -c <Chunk size> -T <vg_name>/<thin-pool-name>
+    ```
+    where:
+    - Size is `LVMClusterSpec.Storage.DeviceClass.ThinPoolConfig.SizePercent`
+    - chunk size is 512KiB, which is the default.
+
+- Manager should also update the `lvmd.yaml` file to map volume group and its thin-pool
```
Comment (nbalacha) [inline doc/design/thin_pool.md:86]:
```suggestion
- VG manager will also update the `lvmd.yaml` file to map volume group and its thin-pool to the topolvm device class
```

File: doc/design/thin_pool.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+### Monitoring and Alerts
+- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
+- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
+- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
+
+
+### Open questions
+- What should be the chunk size of the thin-pools?
+    - Use default size a 512 kiB for now.
+- Should LVMO notify changes of thin-pool if it happens to topolvm?
```
Comment (nbalacha) [inline doc/design/thin_pool.md:108]: Please clarify this point. What changes are you referring to ?

File: doc/design/thin_pool.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+- How should vg-manager handle expansion of thin-pool in near future?
+- After firing alerts should LVMO take any pre-cautionary steps to make thin-lvs occupying whole thin-pool without effecting snapshots
```
Comment (nbalacha) [inline doc/design/thin_pool.md:110]: Please clarify this point as it is unclear.

File: doc/design/thin_pool.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
+# LVMO: Thin provisioning
+
+## Summary
+Allow LVM operator to create thin-pools along with the volume groups.
```
Comment (nbalacha) [inline doc/design/thin_pool.md:4]:
```suggestion
LVM thin provisioning allows creation of volumes whose combined size is greater than that of the available storage. 
Advantages: This leads to better storage utilization etc etc
Disadvantages : etc etc

LVM does this by allocating the blocks in a thin LV from a special "thin pool LV". A thin pool LV must be created before thin LVs can be created within it.

The LVMO will create a thin-pool LV in the volume group in order to create thinly provisioned volumes.
```

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: doc/design/thin_pool.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+- Reduced performance when compared to thick volumes.
+- Over-allocation of the space. This can be mitigated by better monitoring of the disk usage.
+
+LVM does this by allocating the blocks in a thin LV from a special "thin pool LV". A thin pool LV must be created before thin LVs can be created within it.
+
+The LVMO will create a thin-pool LV in the volume group in order to create thinly provisioned volumes.
+
+
+## Proposal:
+- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a volume group and a thin-pool.
```
Comment (leelavg) [inline doc/design/thin_pool.md:20]:
```suggestion
- The `deviceClass` API in the `LVMClusterSpec` will contain the mapping between a device-class and a thin-pool in volume group.
```

File: doc/design/thin_pool.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+- One volume group will be mapped to a single thin pool.
```
Comment (leelavg) [inline doc/design/thin_pool.md:21]:
```suggestion
- One device-class will be mapped to a single thin pool.
```

File: doc/design/thin_pool.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
++   // SizePercent represents percentage of remaining space in the volume group that should be used
++	// for creating the thin pool.
++	// +kubebuilder:validation:default=75
++	// +kubebuilder:validation:Minimum=10
+	// +kubebuilder:validation:Maximum=90
++	SizePercent int `json:"sizePercent,omitempty"`
+
++	// OverProvisionRatio represents the ratio of overprovision that can
++	// be allowed on thin pools
++	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
```
Comment (leelavg) [inline doc/design/thin_pool.md:49]: - this is still in question, it'll definitely be `int` need to know whether it represents percentage or multiplier
- either of the case, better to have `+kubebuilder:validation:Minimum`

File: doc/design/thin_pool.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+type DeviceClass struct {
+    Name string `json:"name,omitempty"`
+
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig  ThinPoolConfig `json:"thinPoolConfig,omitempty"`
```
Comment (leelavg) [inline doc/design/thin_pool.md:59]: - maybe use a pointer which helps in checking against `nil` for absence and drop the omitemtpy as well.

File: doc/design/thin_pool.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    // included in this VolumeGroup
+	// +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+
+    // NodeSelector chooses nodes
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+
++   // ThinPoolConfig contains configurations for the thin-pool
++   ThinPoolConfig ThinPoolConfig `json:"thinPoolConfig,omitempty"`
```
Comment (leelavg) [inline doc/design/thin_pool.md:84]: - same as above

File: doc/design/thin_pool.md (hunk 5 above)
Comment (leelavg) [reply doc/design/thin_pool.md:120]: - this question can be dropped, topolvm by default updates annotations in 10 min intervals if no ops performed in that time space

File: doc/design/thin_pool.md (hunk 6 above)
Comment (leelavg) [reply doc/design/thin_pool.md:122]: - the original intention was, when thinpool gets fulled snapshots are dropped and as I wasn't able to recall where I read that this question can be dropped as well.

File: doc/design/thin_pool.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
+- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
+- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
+
+
+### Open questions
+- What should be the chunk size of the thin-pools?
+    - Use default size a 512 kiB for now.
+- Should LVMO notify changes of thin-pool if it happens to topolvm?
+- How should vg-manager handle expansion of thin-pool in near future?
```
Comment (leelavg) [inline doc/design/thin_pool.md:121]: - as we are not tracking state of VG or thinpool and lvmo doesn't get effected by these changes this question can be dropped

File: doc/design/thin_pool.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+- After firing alerts should LVMO take any pre-cautionary steps to make thin-lvs occupying whole thin-pool without effecting snapshots
+- Do we want to leverage dm-eventd for auto-expansion of thinpool, how could Topolvm monitoring and dm-eventd be made to work in tandem
```
Comment (leelavg) [inline doc/design/thin_pool.md:123]: - as we'll be using topolvm solely for monitoring, no need for dm-eventd for auto-expansion, this question can be dropped

File: doc/design/thin_pool.md (hunk 6 above)
Comment (sp98) [reply doc/design/thin_pool.md]: dropped all the unanswered questions like suggested.

File: doc/design/thin_pool.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    spare-gb: 10
+    type: thin
+    thin-pool-config:
+        name: pool0
+        overprovision-ratio: 50.0
+```
+
+### Monitoring and Alerts
+- Available thin-pool size (both data and metadata) should be provided by topolvm as prometheus metrics.
+- Threshold limits for the thin-pool should be provide as static values in the prometheus yaml file.
```
Comment (nbalacha) [inline doc/design/thin_pool.md:120]:
```suggestion
- Threshold limits for the thin-pool should be provide as static values in the PrometheusRule.
```

File: doc/design/thin_pool.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+- If available size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
```
Comment (nbalacha) [inline doc/design/thin_pool.md:121]:
```suggestion
- If used size of data or metadata for a particular thin-pool crosses the threshold, then appropriate alerts should be triggered.
```

**Review by nbalacha — APPROVED** [review]

**Pull request merged**